evm/src/get_challenges.rs

```diff
@@ -12,6 +13,117 @@ use crate::permutation::{
     get_n_grand_product_challenge_sets_target,
 };
 use crate::proof::*;
+use crate::util::u256_limbs;
+
+fn observe_root<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+    challenger: &mut Challenger<F, C::Hasher>,
+    root: H256,
+) {
+    for limb in root.into_uint().0.into_iter() {
+        challenger.observe_element(F::from_canonical_u32(limb as u32));
+        challenger.observe_element(F::from_canonical_u32((limb >> 32) as u32));
+    }
+}
+
+fn observe_trie_roots<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+    challenger: &mut Challenger<F, C::Hasher>,
+    trie_roots: &TrieRoots,
+) {
+    observe_root::<F, C, D>(challenger, trie_roots.state_root);
+    observe_root::<F, C, D>(challenger, trie_roots.transactions_root);
+    observe_root::<F, C, D>(challenger, trie_roots.receipts_root);
+}
+
+fn observe_trie_roots_target<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    challenger: &mut RecursiveChallenger<F, C::Hasher, D>,
+    trie_roots: &TrieRootsTarget,
+) where
+    C::Hasher: AlgebraicHasher<F>,
+{
+    challenger.observe_elements(&trie_roots.state_root);
+    challenger.observe_elements(&trie_roots.transactions_root);
+    challenger.observe_elements(&trie_roots.receipts_root);
+}
+
+fn observe_block_metadata<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    challenger: &mut Challenger<F, C::Hasher>,
+    block_metadata: &BlockMetadata,
+) {
+    challenger.observe_elements(
+        &u256_limbs::<F>(U256::from_big_endian(&block_metadata.block_beneficiary.0))[..5],
+    );
+    challenger.observe_element(F::from_canonical_u32(
+        block_metadata.block_timestamp.as_u32(),
+    ));
+    challenger.observe_element(F::from_canonical_u32(block_metadata.block_number.as_u32()));
+    challenger.observe_element(F::from_canonical_u32(
+        block_metadata.block_difficulty.as_u32(),
+    ));
+    challenger.observe_element(F::from_canonical_u32(
+        block_metadata.block_gaslimit.as_u32(),
+    ));
+    challenger.observe_element(F::from_canonical_u32(
+        block_metadata.block_chain_id.as_u32(),
+    ));
+    challenger.observe_element(F::from_canonical_u32(
+        block_metadata.block_base_fee.as_u32(),
+    ));
+}
```

Comment:
Not exactly related to this PR, but I think the base fee often is larger than 2^32, which would lead to a panic here and elsewhere. 
For example, at the time of writing this comment, the [latest block](https://etherscan.io/block/17947828) uses `block_base_fee ~= 13.15 Gwei > 2^32 wei`.

Reply:
Yeah that's true, but by design we only store the Target versions as a single `Target`, hence limited to 32bits for all block metadata. I had written purposely a conversion from `canonical_u32` in `prove_root` to highlight this design limitation. Should we adapt those metadata fields on a case-by-base basis in a follow-up PR?

Reply:
> I had written purposely a conversion from `canonical_u32` in `prove_root` to highlight this design limitation.

I'm not sure I understand what you mean by that.

> Should we adapt those metadata fields on a case-by-base basis in a follow-up PR?

Yes that'd be great. In general, I don't think these metadata field should be `U256` if they're required to fit in a `u32`. 

This PR is good to merge, just noticed this issue while reviewing but it's unrelated to the PR's changes.

Reply:
Ah sorry for the confusion, what I meant is that previously we were converting the block metadata to u64s before setting the target, which was then causing errors in the recursive circuit if they were indeed greater than `u32::MAX`. [This PR](https://github.com/mir-protocol/plonky2/pull/1161) addressed it by converting to u32s to be consistent with the limb size that is expected to fit in a Target.

As for which metadata to have fitting in more than 1 limb, do you see any other field than basefee that would require such a change? The other most likely one seems to be the gaslimit, and as we had discussed this will never be practically relevant.

Reply:
> As for which metadata to have fitting in more than 1 limb, do you see any other field than basefee that would require such a change?

Yes I think basefee is the only field that can be larger than 32-bit (at least for the next 83 years, after which the timestamp would become a problem). 

We should probably open an issue for this, but the block difficulty used to be larger than 32-bit but is now 0 after the Merge. The difficulty has been replaced by prevrandao, so maybe we should just store the prevrandao value in `BlockMetadata.difficulty` or rename it to `BlockMetadata.difficulty_or_prevrandao` or `BlockMetadata.prevrandao`.

Reply:
> Yes I think basefee is the only field that can be larger than 32-bit
Alright, I'll deal with it.

> We should probably open an issue for this, but the block difficulty used to be larger than 32-bit but is now 0 after the Merge.
Are there any plans to deal with Ethereum state pre Merge? Otherwise I don't think that'd be a big issue.

Reply:
> Are there any plans to deal with Ethereum state pre Merge? Otherwise I don't think that'd be a big issue.

I don't think so. It's just that if we put the prevrandao value in `BlockMetadata.difficulty`, then we'd need to use 8 limbs for `BlockMetadataTarget.difficulty`.